### PR TITLE
Use BaseTypeInfo API to fix up polymorphic schemas

### PR DESCRIPTION
**Pull request by captainsafia**
This PR is a follow-up to the original introduction of support for polymorphic schemas in https://github.com/dotnet/aspnetcore/pull/56466 and reacts to APIs introduced in https://github.com/dotnet/runtime/issues/104046. This PR uses the new `JsonSchemaExporterContext.BaseTypeInfo` property to apply to changes:

For polymorphic types where the base type is a non-abstract class without a discriminator, System.Text.Json will support serializing to that base type when no discriminator is provided. As a result, the base type will be included as one of the possible subschemas that a polymorphic schema can match to in the `anyOf` property. OpenAPI requires that _all_ polymorphic subschemas have a type discriminator associated with them to support round-tripping. ~~To address this mismatch in schema implementations, we mark sub-schemas associated with non-abstract base types that lack a discirminator and remove them from the `anyOf` subschemas list when converting the types.~~ After some discussion with @mikekistler, to maintain accuracy with the actual serialization happening, when we run into types that match this constraint we'll instead emit an `anyOf` schema _without_ the `discriminator` mapping.

Prior to the existence of `JsonSchemaExporterContext.BaseTypeInfo`, we used a less reliable workaround (`context.Path`) to check if we were generating a subschema. Here, we use the new property to check if we are in subschema or not and apply fixes to make sure that all `anyOf` subschemas are referenced correctly.

### CODE REVIEW

**Comment by captainsafia** [comment]
Moving this back into draft mode while I work on some updates to this PR.

@mikekistler shared some insights on an alternative way that we can represent "polymorphic types with a non-abstract base class that has no discriminator" in the schema that will be more accurate and comply with OpenAPI's constraints.

**Review by mikekistler — APPROVED** [review]
Looks good! 👍

File: src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs [hunk 1]
```diff
@@ -327,10 +327,19 @@ internal static void ApplyParameterInfo(this JsonNode schema, ApiParameterDescri
     /// <param name="createSchemaReferenceId">A delegate that generates the reference ID to create for a type.</param>
     internal static void ApplyPolymorphismOptions(this JsonNode schema, JsonSchemaExporterContext context, Func<JsonTypeInfo, string?> createSchemaReferenceId)
     {
-        // The `context.Path.Length == 0` check is used to ensure that we only apply the polymorphism options
+        // The `context.BaseTypeInfo == null` check is used to ensure that we only apply the polymorphism options
```
Comment (amcasey) [inline src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:330]: Does schema nesting represent subtyping?  I would have guessed it was for composition.  Why would having no base type (I'm assuming there's no "top" type like `object`) correspond to having no containing schema?

File: src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs [hunk 2]
```diff
@@ -327,10 +327,19 @@ internal static void ApplyParameterInfo(this JsonNode schema, ApiParameterDescri
...
         // to the top-level schema and not to any nested schemas that are generated.
-        if (context.TypeInfo.PolymorphismOptions is { } polymorphismOptions && context.Path.Length == 0)
+        if (context.TypeInfo.PolymorphismOptions is { } polymorphismOptions && context.BaseTypeInfo == null)
         {
+            // System.Text.Json supports serializing to a non-abstract base class if no discriminator is provided.
+            // OpenAPI requires that all polymorphic sub-schemas have an associated discriminator. To maintain accuracy,
+            // with the actual implementation and comply with the limitations of the OpenAPI schema we do not emit
```
Comment (amcasey) [inline src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:336]: The "actual implementation" being STJ?

File: src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs [hunk 3]
```diff
@@ -327,10 +327,19 @@ internal static void ApplyParameterInfo(this JsonNode schema, ApiParameterDescri
...
+            // a `discriminator` property for types that are represented this way.
+            if (!context.TypeInfo.Type.IsAbstract &&
+                !polymorphismOptions.DerivedTypes.Any(type => type.DerivedType == context.TypeInfo.Type))
```
Comment (amcasey) [inline src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:339]: I'm guessing `polymorphismOptions` includes an exhaustive list of base-type/subtype pairs?

Comment (amcasey) [inline src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:339]: Isn't the interesting case when the non-abstract class is the base type?  I might be mixed up.  Why are we checking whether it's the derived type?

File: src/OpenApi/src/Schemas/OpenApiJsonSchema.Helpers.cs [hunk 4]
```diff
@@ -300,7 +300,10 @@ public static void ReadProperty(ref Utf8JsonReader reader, string propertyName,
             case OpenApiSchemaKeywords.DiscriminatorMappingKeyword:
                 reader.Read();
                 var mappings = ReadDictionary<string>(ref reader);
-                schema.Discriminator.Mapping = mappings;
+                if (mappings is not null)
```
Comment (amcasey) [inline src/OpenApi/src/Schemas/OpenApiJsonSchema.Helpers.cs:303]: Is this just an unrelated bug fix?

File: src/OpenApi/src/Services/Schemas/OpenApiSchemaStore.cs [hunk 5]
```diff
@@ -112,13 +121,32 @@ public void PopulateSchemaIntoReferenceCache(OpenApiSchema schema, bool captureS
                 AddOrUpdateSchemaByReference(anyOfSchema, baseTypeSchemaId);
             }
         }
-        if (schema.Properties is not null)
+
+        if (schema.Items is not null)
+        {
+            AddOrUpdateAnyOfSubSchemaByReference(schema.Items);
```
Comment (amcasey) [inline src/OpenApi/src/Services/Schemas/OpenApiSchemaStore.cs:127]: Was this already recursive?

Comment (amcasey) [inline src/OpenApi/src/Services/Schemas/OpenApiSchemaStore.cs:127]: I'm guessing we're relying on the C# type system to protect us from cycles?

File: src/OpenApi/test/Microsoft.AspNetCore.OpenApi.Tests/Shared/SharedTypes.Polymorphism.cs [hunk 6]
```diff
@@ -72,3 +72,20 @@ internal class FabricColor : Color
 {
     public required string Dye { get; set; }
 }
+
+[JsonDerivedType(typeof(Cat), typeDiscriminator: "cat")]
+[JsonDerivedType(typeof(Dog), typeDiscriminator: "dog")]
+[JsonDerivedType(typeof(Pet), typeDiscriminator: "pet")]
+internal class Pet
```
Comment (amcasey) [inline src/OpenApi/test/Microsoft.AspNetCore.OpenApi.Tests/Shared/SharedTypes.Polymorphism.cs:79]: This is in contrast to `Person` where the base type is abstract?  If so, a comment explaining the significance of the structure might be helpful.

**Review by amcasey — COMMENTED** [review]
I added some comments, but I probably need an offline overview.

File: src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs (hunk 1 above)
Comment (captainsafia) [reply src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:330]: `BaseTypeInfo` is strictly related with polymorphic subtyping in this case (see the implementation of the API [here](https://github.com/dotnet/runtime/commit/7f5ce1e2bb5b7589281a713dc379671cc0b99ea6) for some additional insight). In this case, the `BaseTypeInfo` is only set if we're generating the subschema for a derived type.

File: src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs (hunk 2 above)
Comment (captainsafia) [reply src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:336]: Yep, actual implementation here means what the STJ serializers do when the app is running.

File: src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs (hunk 3 above)
Comment (captainsafia) [reply src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:339]: It's exhaustive in the sense that it matches what the serialier uses to determine derived types for polymorphic schemas. System.Text.Json populates this list based on the [JsonDerivedType attributes](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/polymorphism?pivots=dotnet-8-0#serialize-properties-of-derived-classes) on the base type.

Comment (captainsafia) [reply src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:339]: `PolymorphismOptions.DerivedTypes` is populated from the `JsonDerivedTypeAttribute` annotations on a type, which provide a way for defining what discriminator is assocaited with a type.  We care when there is no derived type assocated with the base type because then there is no discriminator defined.

Essentially, we're checking for the difference between:

```
// Scenario A:
[JsonDerivedType(typeof(Cat), typeDiscriminator: "cat")]
[JsonDerivedType(typeof(Dog), typeDiscriminator: "dog")]
[JsonDerivedType(typeof(Pet), typeDiscriminator: "pet")]
internal class Pet { } 
```

and

```
// Scenario B:
[JsonDerivedType(typeof(Cat), typeDiscriminator: "cat")]
[JsonDerivedType(typeof(Dog), typeDiscriminator: "dog")]
internal class Pet { } 
```

In Scenario A, there's a discriminator associated with the `Pet` type so we can use it in compliance with OpenAPI's expectations. In Scenario B, there's no discriminator assocaited with `Pet` type but STJ will still  deserialize to it if no `$type` property is provided in a JSON object.

File: src/OpenApi/test/Microsoft.AspNetCore.OpenApi.Tests/Shared/SharedTypes.Polymorphism.cs (hunk 6 above)
Comment (captainsafia) [reply src/OpenApi/test/Microsoft.AspNetCore.OpenApi.Tests/Shared/SharedTypes.Polymorphism.cs:79]: This is moreso in contrast to `Color` where the base type is non-abstract _and_ not defined in the derived types list, per the explanation in [this comment](https://github.com/dotnet/aspnetcore/pull/56908/files/998eff1f86397dca153f087921b32ac2db8ab464#r1696003064).

**Comment by amcasey** [comment]
> For polymorphic types where the base type is a non-abstract class without a discriminator, System.Text.Json will support serializing to that base type when no discriminator is provided.

To add some context for other newbies and to check my understanding:

STJ will allow you to serialize a `Derived` as a `Base` without a discriminator property and `Derived`'s properties will end up in the resulting JSON but the JSON will deserialize back to a `Base` (not a `Derived`) _without_ the `Derived` properties.  So the "round-trip" is lossy.  If you want to get back what you started with (an object with runtime type `Derived`), you need a discriminator property.

File: src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs (hunk 3 above)
Comment (amcasey) [reply src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:339]: Okay, I was pretty confused by the name of the `DerivedType` property.  In this case `type` is a `JsonDerivedType`, which is effectively a type info/wrapper and `type.DerivedType` is the wrapped type.  That is, `DerivedType` is the type wrapped by the `JsonDerivedType` and not a type derived from it.

Comment (amcasey) [reply src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:339]: (I had originally thought each element of `DerivedTypes` probably contained both `BaseType` and `DerivedType` properties and was based on the type system, rather than attributes on the base type).

File: src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs (hunk 1 above)
Comment (amcasey) [reply src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:330]: "Subschema" in the sense that it's one of several in a discriminated union?

File: src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs (hunk 2 above)
Comment (amcasey) [reply src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:336]: I would find it clearer to say "STJ" then.

File: src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs (hunk 3 above)
Comment (amcasey) [reply src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:339]: "If the base type doesn't declare itself as its own derived type, then it can't have specified a discriminator, which OpenAPI requires."?

File: src/OpenApi/test/Microsoft.AspNetCore.OpenApi.Tests/Shared/SharedTypes.Polymorphism.cs (hunk 6 above)
Comment (amcasey) [reply src/OpenApi/test/Microsoft.AspNetCore.OpenApi.Tests/Shared/SharedTypes.Polymorphism.cs:79]: I think it might be helpful if there were comments describing the salient features of each type hierarchy.

File: src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs [hunk 7]
```diff
@@ -327,10 +327,19 @@ internal static void ApplyParameterInfo(this JsonNode schema, ApiParameterDescri
...
+            {
+                return;
```
Comment (amcasey) [inline src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:341]: I'm reading this early return as "don't apply the polymorphism options", but I don't fully understand the consequences of that.  There's some explanation in your PR description, but it might be helpful to incorporate some of that into the code.

I think maybe "ApplyPolymorphismOptions" means "include discriminator properties" and the early return is a way of saying "don't include discriminator properties", which will have the effect of emitting "any" (which I think is an undiscriminated union, rather than a wildcard).  Is that right?

**Review by amcasey — COMMENTED** [review]
I'm getting closer, but I still have more questions.

File: src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs (hunk 2 above)
Comment (captainsafia) [reply src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs]: Updated.

File: src/OpenApi/test/Microsoft.AspNetCore.OpenApi.Tests/Shared/SharedTypes.Polymorphism.cs (hunk 6 above)
Comment (captainsafia) [reply src/OpenApi/test/Microsoft.AspNetCore.OpenApi.Tests/Shared/SharedTypes.Polymorphism.cs:88]: Updated.

File: src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs (hunk 7 above)
Comment (captainsafia) [reply src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:343]: > I think maybe "ApplyPolymorphismOptions" means "include discriminator properties" and the early return is a way of saying "don't include discriminator properties", which will have the effect of emitting "any" (which I think is an undiscriminated union, rather than a wildcard). Is that right?

Yep. I've updated the name of the method to be a little be clearer about how the polymorphism options are mapped to the OpenAPI document.

File: src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs (hunk 3 above)
Comment (captainsafia) [reply src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:341]: Updated.

File: src/OpenApi/src/Services/Schemas/OpenApiSchemaStore.cs (hunk 5 above)
Comment (captainsafia) [reply src/OpenApi/src/Services/Schemas/OpenApiSchemaStore.cs:127]: It wasn't recursive before and FWIW it's not entirely recursive now.

The changes related to `AddOrUpdateAnyOfSubSchemaByReference` are specifically addressing the scenario where a polymorphic type appears as part of another type (e.g. `List<Shape>` or `Shape` as a property in a container type). The set of tests [included here](https://github.com/dotnet/aspnetcore/blob/safia/openapi-fixes/src/OpenApi/test/Microsoft.AspNetCore.OpenApi.Tests/Transformers/SchemaTransformerTests.cs#L550) cover this scenario.

File: src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs (hunk 1 above)
Comment (amcasey) [reply src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:331]: As discussed offline, type hierarchies are represented as discriminated unions and subtypes are subschemas in the sense that they are nested within the discriminated union schema.

File: src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs [hunk 8]
```diff
@@ -320,21 +320,23 @@ internal static void ApplyParameterInfo(this JsonNode schema, ApiParameterDescri
...
         // to the top-level schema and not to any nested schemas that are generated.
         if (context.TypeInfo.PolymorphismOptions is { } polymorphismOptions && context.BaseTypeInfo == null)
         {
             // System.Text.Json supports serializing to a non-abstract base class if no discriminator is provided.
-            // OpenAPI requires that all polymorphic sub-schemas have an associated discriminator. To maintain accuracy,
-            // with the actual implementation and comply with the limitations of the OpenAPI schema we do not emit
-            // a `discriminator` property for types that are represented this way.
+            // OpenAPI requires that all polymorphic sub-schemas have an associated discriminator. If the base type
+            // doesn't declare itself as its own derived type via [JsonDerived], then it can't have a discriminator,
+            // which OpenAPI requires. In that case, we don't exit early to avoid mapping the polymorphism options
```
Comment (amcasey) [inline src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:338]:
```suggestion
            // which OpenAPI requires. In that case, we exit early to avoid mapping the polymorphism options
```

Comment (amcasey) [reply src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:338]: I think there might be an extra negative, but I'm not sure.

**Review by amcasey — COMMENTED** [review]
Love the new comments

File: src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs [hunk 9]
```diff
@@ -362,6 +373,14 @@ internal static void ApplySchemaReferenceId(this JsonNode schema, JsonSchemaExpo
         {
             schema[OpenApiConstants.SchemaId] = schemaReferenceId;
         }
+        // If the type is a non-abstract base class that is not one of the derived types then mark it as a base schema.
+        if (context.TypeInfo.PolymorphismOptions is { } polymorphismOptions &&
+            context.BaseTypeInfo == context.TypeInfo &&
```
Comment (amcasey) [inline src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:378]: Could the code for this check be shared?  It seems like it would be easy for the two implementations to get slightly out of sync.

File: src/OpenApi/src/Services/Schemas/OpenApiSchemaStore.cs (hunk 5 above)
Comment (amcasey) [reply src/OpenApi/src/Services/Schemas/OpenApiSchemaStore.cs:127]: So we're going to depth-first-search traverse the schema graph until we find an `AnyOf` and then go back to the old mechanism (`AddOrUpdateSchemaByReference`)?

**Review by amcasey — COMMENTED** [review]
Do we have a test for schematizing a self-referential type like a linked list or a tree?

**Comment by captainsafia** [comment]
> Do we have a test for schematizing a self-referential type like a linked list or a tree?

https://github.com/dotnet/aspnetcore/blob/b4558f77c6da96cc58f54618bfd104246beb6328/src/OpenApi/test/Microsoft.AspNetCore.OpenApi.Tests/Services/OpenApiSchemaService/OpenApiSchemaService.RequestBodySchemas.cs#L199-L201 covers this with a type defined as follows:

```csharp
internal class Proposal
{
    public required Proposal ProposalElement { get; set; }
    public required Stream Stream { get; set; }
}
```

File: src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs (hunk 9 above)
Comment (captainsafia) [reply src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs:378]: It can be, it's admittedly difficult to come up with a good method name for the shared logic but I think I did decent enough.

**Review by amcasey — APPROVED** [review]

**Comment by captainsafia** [comment]
## Support calling `ProducesProblem` and `ProducesValidationProblem` on route groups

The `ProducesProblem` and `ProducesValidationProblem` extension methods have been updated to support application on route groups, as in the sample below. These methods can be used to indicate that all endpoints in a route group can return `ProblemDetails` or `ValidationProblemDetails` responses for the purposes of OpenAPI metadata.

```csharp
var app = WebApplication.Create();

var todos = app.MapGroup("/todos")
	.ProducesProblem();

todos.MapGet("/", () => new Todo(1, "Create sample app", false));
todos.MapPost("/", (Todo todo) => Results.Ok(todo));

app.Run();

record Todo(int Id, string Title, boolean IsCompleted);
```

## Microsoft.AspNetCore.OpenApi is trim and native AoT-compatible

The new built-in OpenAPI support in ASP.NET Core is trim and native AoT-compatible for use in applications that are published with AoT enabled.

```csharp
var builder = WebApplication.CreateBuilder(args);

builder.Services.AddOpenApi();

var app = builder.Build();

// Configure the HTTP request pipeline.
if (app.Environment.IsDevelopment())
{
    app.MapOpenApi();
}

app.UseHttpsRedirection();

var summaries = new[]
{
    "Freezing", "Bracing", "Chilly", "Cool", "Mild", "Warm", "Balmy", "Hot", "Sweltering", "Scorching"
};

app.MapGet("/weatherforecast", () =>
{
    var forecast =  Enumerable.Range(1, 5).Select(index =>
        new WeatherForecast
        (
            DateOnly.FromDateTime(DateTime.Now.AddDays(index)),
            Random.Shared.Next(-20, 55),
            summaries[Random.Shared.Next(summaries.Length)]
        ))
        .ToArray();
    return forecast;
})
.WithName("GetWeatherForecast");

app.Run();

record WeatherForecast(DateOnly Date, int TemperatureC, string? Summary)
{
    public int TemperatureF => 32 + (int)(TemperatureC / 0.5556);
}
```

When the application above is published, no warnings are produced.

```
$ dotnet publish /p:PublishAot=true 
Restore complete (0.4s)
You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
  publish-aot-test succeeded (12.6s) → bin/Release/net9.0/osx-arm64/publish/

Build succeeded in 13.2s
```

Note: that as of 9.0-preview7, an explicit reference to v1.6.16 of `Microsoft.OpenApi` is required in the target project to fully support native AoT.

```diff
<ItemGroup>
+    <PackageReference Include="Microsoft.OpenApi" Version="1.6.16" />
</ItemGroup>
```

## Improvements to transformer registration APIs in Microsoft.AspNetCore.OpenApi

OpenAPI transformers support modifying the OpenAPI document, operations within the document, or schemas associated with types in the API. In this preview, the APIs for registering transformers on an OpenAPI document to provide a variety of options for registering transformers.

Previously, the following APIs where available for registering transformers:

```csharp
OpenApiOptions UseTransformer(Func<OpenApiDocument, OpenApiDocumentTransformerContext, CancellationToken> transformer)
OpenApiOptions UseTransformer(IOpenApiDocumentTransformer transformer)
OpenApiOptions UseTransformer<IOpenApiDocumentTransformer>()
OpenApiOptions UseSchemaTransformer(Func<OpenApiSchema, OpenApiSchemaTransformerContext, CancellationToken>)
OpenApiOptions UseOperationTransformer(Func<OpenApiOperation, OpenApiOperationTransformerContext, CancellationToken>)
```

The new API set is as follows:

```csharp
OpenApiOptions AddDocumentTransformer(Func<OpenApiDocument, OpenApiDocumentTransformerContext, CancellationToken> transformer)
OpenApiOptions AddDocumentTransformer(IOpenApiDocumentTransformer transformer)
OpenApiOptions AddDocumentTransformer<IOpenApiDocumentTransformer>()

OpenApiOptions AddSchemaTransformer(Func<OpenApiSchema, OpenApiSchemaTransformerContext, CancellationToken> transformer)
OpenApiOptions AddSchemaTransformer(IOpenApiSchemaTransformer transformer)
OpenApiOptions AddSchemaTransformer<IOpenApiSchemaTransformer>()

OpenApiOptions AddOperationTransformer(Func<OpenApiOperation, OpenApiOperationTransformerContext, CancellationToken> transformer)
OpenApiOptions AddOperationTransformer(IOpenApiOperationTransformer transformer)
OpenApiOptions AddOperationTransformer<IOpenApiOperationTransformer>()
```

Thanks to GitHub user "martincostello" for contributing this API improvement to the package!
